File: src/MutableMap/MutableMap.java
/*******************************************************************************
*  Copyright by the contributors to the Dafny Project
*  SPDX-License-Identifier: MIT
*******************************************************************************/
package DafnyLibraries;

import dafny.DafnySet;
import dafny.DafnyMap;
import dafny.Tuple2;
import dafny.TypeDescriptor;
import dafny.Helpers;

import java.util.concurrent.*;
import java.util.ArrayList;
import java.util.Map;
import java.math.BigInteger;

public class MutableMap<K,V> extends _ExternBase_MutableMap<K,V> {
  private ConcurrentHashMap<K,V> m;

  public MutableMap(dafny.TypeDescriptor<K> _td_K, dafny.TypeDescriptor<V> _td_V) {
    super(_td_K, _td_V);
    m = new ConcurrentHashMap<K,V>();
  }

  @Override 
  public DafnyMap<K,V> content() {
    return new DafnyMap<K,V>(m);
  }

  @Override 
  public void Put(K k, V v) {
    m.put(k, v);
  }

  @Override 
  public DafnySet<? extends K> Keys() {
    return new DafnySet(m.keySet());
  }

  @Override 
  public boolean HasKey(K k) {
    return m.containsKey(k);
  }

  @Override 
  public DafnySet<? extends V> Values() {
    return new DafnySet(m.values());
  }

  @Override 
  public DafnySet<? extends Tuple2<K,V>> Items() {
    ArrayList<Tuple2<K, V>> list = new ArrayList<Tuple2<K, V>>();
    for (Map.Entry<K, V> entry : m.entrySet()) {
      list.add(new Tuple2<K, V>(entry.getKey(), entry.getValue()));
    }
    return new DafnySet<Tuple2<K, V>>(list);
  }

  @Override 
  public V Select(K k) {
    return m.get(k);
  }

  @Override 
  public void Remove(K k) {
    m.remove(k);
  }

  @Override 
  public BigInteger Size() {
    return BigInteger.valueOf(m.size());
  }
}

File: src/MutableMap/MutableMap.cs
/*******************************************************************************
*  Copyright by the contributors to the Dafny Project
*  SPDX-License-Identifier: MIT
*******************************************************************************/
using System.Collections;
using System.Collections.Concurrent;
using System.Collections.Immutable;
using System.Numerics;
using _System;
using Dafny;


namespace DafnyLibraries {
   public partial class MutableMap<K, V>
   {
      private ConcurrentDictionary<K, V> m;
      public MutableMap() {
         m = new ConcurrentDictionary<K, V>();
      }

      public IMap<K, V> content() {
         var keyPairs = new List<IPair<K, V>>();
         foreach (var k in m.Keys) {
            keyPairs.Add(new Pair<K, V>(k, m[k]));
         }
         return Map<K, V>.FromCollection(keyPairs);
      }

      public void Put(K k, V v)
      {
         m[k] = v;
      }

      public Dafny.ISet<K> Keys()
      {
         var keys = m.Keys;
         return Set<K>.FromCollection(keys);
      }

      public bool HasKey(K k)
      {
         return m.ContainsKey(k);
      }

      public Dafny.ISet<V> Values()
      {
         var values= m.Values;
         return Set<V>.FromCollection(values);
      }

      public Dafny.ISet<_ITuple2<K, V>> Items()
      {
         var keyPairs = new List<_ITuple2<K, V>>();
         foreach (var k in m.Keys) {
            keyPairs.Add(new Tuple2<K, V>(k, m[k]));
         }
         return Set<_ITuple2<K,V>>.FromCollection(keyPairs);
      }

      public V Select(K k)
      {
         return m[k];
      }

      public void Remove(K k)
      {
         if (HasKey(k)) {
            var keypair = new KeyValuePair<K, V>(k, m[k]); 
            m.TryRemove(keypair);
         }
      }

      public BigInteger Size()
      {
         return new BigInteger(m.Count);
      }
   }
}
